### PR TITLE
Support nested records in the Arrow Builder

### PR DESCRIPTION
**Pull request by dominiklohmann**
###  :notebook_with_decorative_cover: Description

This adds support for records to the Arrow table slice builder.

###  :memo: Checklist

- [x] All user-facing changes have changelog entries.
- [x] The changes are reflected on [docs.tenzir.com/vast](https://docs.tenzir.com/vast), if necessary.
- [x] The PR description contains instructions for the reviewer, if necessary.

### :dart: Review Instructions

File-by-file.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -13,6 +13,12 @@ This changelog documents all notable user-facing changes of VAST.
 
 ## Unreleased
 
+- 🐞 VAST now supports nested records in both Arrow- and MessagePack-encoded
+  table slices, e.g., data of type `list<record<name: string, age:
+  count>`. While not yet support in all readers, ingesting such data will no
+  longer cause VAST to crash.
```
Comment (mavam) [inline CHANGELOG.md:19]:
```suggestion
- 🐞 VAST now supports nested records in both Arrow and MessagePack
  table slices, e.g., data of type `list<record<name: string, age: count>`.
  While not yet queryable and supported in all formats, ingesting such data will no
  longer cause VAST to crash.
```

File: libvast/src/arrow_table_slice_builder.cpp [hunk 2]
```diff
@@ -388,6 +387,63 @@ class map_column_builder : public arrow_table_slice_builder::column_builder {
   std::unique_ptr<column_builder> val_builder_;
 };
 
+class record_column_builder : public arrow_table_slice_builder::column_builder {
+public:
+  // There is no RecordBuilder in Arrow. A record is simply a list of structs
+  // (key-value pairs).
+  using arrow_builder_type = arrow::ListBuilder;
+
```
Comment (mavam) [inline libvast/src/arrow_table_slice_builder.cpp:395]:
```suggestion
```

File: libvast/src/arrow_table_slice_builder.cpp [hunk 3]
```diff
@@ -388,6 +387,63 @@ class map_column_builder : public arrow_table_slice_builder::column_builder {
...
+    return result;
+  }
+
+  std::shared_ptr<arrow::ArrayBuilder> arrow_builder() const override {
+    return list_builder_;
+  }
+
+private:
+  std::shared_ptr<arrow::StructBuilder> struct_builder_;
+  std::shared_ptr<arrow_builder_type> list_builder_;
```
Comment (mavam) [inline libvast/src/arrow_table_slice_builder.cpp:442]:
```suggestion
  std::shared_ptr<arrow::ListBuilder> list_builder_;
```

Only if `Builder::arrow_builder_type` is not a thing.

**Review by mavam — CHANGES_REQUESTED** [review]
The implementation looks good.

I would like to see a brief unit and/or integration test before greenlighting this PR.

File: libvast/src/format/json.cpp [hunk 4]
```diff
@@ -268,6 +268,26 @@ caf::expected<data> type_biased_convert_impl<::simdjson::dom::object, map>(
...
 }
 
+template <>
+caf::expected<data> type_biased_convert_impl<::simdjson::dom::object, record>(
+  ::simdjson::dom::object o, const type& t) {
+  const auto& r = dynamic_cast<const record_type&>(*t);
+  record xs;
+  xs.reserve(o.size());
+  for (auto [k, v] : o) {
+    if (const auto* rf = r.find(k); !rf) {
```
Comment (mavam) [inline libvast/src/format/json.cpp:278]:
```suggestion
  ::simdjson::dom::object o, const type& t) {
  const auto* r = caf::get_if<const record_type>(&t);
  VAST_ASSERT(r != nullptr);
  record xs;
  xs.reserve(o.size());
  for (auto [k, v] : o) {
    if (const auto* rf = r->find(k); !rf) {
```

Any reason not to do this?

**Review by mavam — COMMENTED** [review]
Why is the second baseline empty? Shouldn't it return the ingested data?

Since we're not querying from the nested record, I thought this would work.

File: libvast/src/format/json.cpp (hunk 4 above)
Comment (dominiklohmann) [reply libvast/src/format/json.cpp:278]: No need to go through sum type access again, we already know at this point that we have a map. Also, if we change anything here, we should change it for all visitors at the same time.

**Comment by dominiklohmann** [comment]
> Why is the second baseline empty? Shouldn't it return the ingested data?

Because it is non-functional at this point. I posted about the issue in Slack and would like to evaluate this with either @tobim or you sometime today—a second pair of eyes may help here.

File: vast/integration/reference/nested-records/step_01.ref [hunk 5]
```diff
@@ -0,0 +1 @@
+<[<"b": "123", "c": 456>], 789>
```
Comment (mavam) [inline vast/integration/reference/nested-records/step_01.ref:1]: Nice, this is what I wanted to see!

File: libvast/src/index/list_index.cpp [hunk 6]
```diff
@@ -67,12 +68,18 @@ bool list_index::append_impl(data_view x, id pos) {
...
+                      detail::pretty_type_name(this), value_type_);
         }
       }
       auto x = v->begin();
-      for (auto i = 0u; i < seq_size; ++i)
-        elements_[i]->append(*x++, pos);
+      for (auto i = 0u; i < seq_size; ++i) {
+        if (elements_[i])
+          elements_[i]->append(*x++, pos);
+      }
```
Comment (mavam) [inline libvast/src/index/list_index.cpp:82]:
```suggestion
      for (auto i = 0u; i < seq_size; ++i)
        if (elements_[i])
          elements_[i]->append(*x++, pos);
```

**Review by mavam — APPROVED** [review]
Nice work. This is perfect prep for the upcoming nested lists change.

I didn't test this locally but verified the integration test baseline matches my expectation.

File: libvast/src/index/list_index.cpp [hunk 7]
```diff
@@ -67,12 +68,18 @@ bool list_index::append_impl(data_view x, id pos) {
         elements_.resize(seq_size);
         for (auto i = old; i < elements_.size(); ++i) {
           elements_[i] = factory<value_index>::make(value_type_, options());
-          VAST_ASSERT(elements_[i]);
+          // TODO: Support the #skip attribute here.
```
Comment (dominiklohmann) [inline libvast/src/index/list_index.cpp:71]: We should write a story for this.

File: libvast/src/arrow_table_slice_builder.cpp [hunk 8]
```diff
@@ -276,16 +275,15 @@ class column_builder_impl final
   std::shared_ptr<arrow_builder_type> arrow_builder_;
 };
 
-template <class SequenceType>
-class sequence_column_builder
-  : public arrow_table_slice_builder::column_builder {
+template <class ListType>
+class list_column_builder : public arrow_table_slice_builder::column_builder {
```
Comment (dominiklohmann) [inline libvast/src/arrow_table_slice_builder.cpp:279]: Remove useless template argument—it's always `list_type`.

File: libvast/src/arrow_table_slice_builder.cpp [hunk 9]
```diff
@@ -418,9 +468,16 @@ arrow_table_slice_builder::column_builder::make(const type& t,
       return std::make_unique<map_column_builder>(pool, make_arrow_type(fields),
                                                   std::move(key_builder),
                                                   std::move(value_builder));
+      // record_type fields{{"key", x.key_type}, {"value", x.value_type}};
+      // return column_builder::make(fields, pool);
```
Comment (dominiklohmann) [inline libvast/src/arrow_table_slice_builder.cpp:472]: Remove.

File: libvast/src/arrow_table_slice.cpp [hunk 10]
```diff
@@ -167,68 +200,71 @@ void decode(const type& t, const arrow::Array& arr, F& f) {
...
     }
     case arrow::Type::UINT32: {
       using array_type = arrow::NumericArray<arrow::UInt32Type>;
-      return decode(t, static_cast<const array_type&>(arr), f);
+      return decode(t, dynamic_cast<const array_type&>(arr), f);
     }
     case arrow::Type::UINT64: {
       using array_type = arrow::NumericArray<arrow::UInt64Type>;
-      return decode(t, static_cast<const array_type&>(arr), f);
+      return decode(t, dynamic_cast<const array_type&>(arr), f);
```
Comment (dominiklohmann) [inline libvast/src/arrow_table_slice.cpp:267]: Revert to `static_cast` and change clang-tidy rules accordingly.

File: vast/integration/default_set.yaml [hunk 11]
```diff
@@ -438,6 +438,15 @@ tests:
       - command: -N export json '#type ni "suricata"'
         transformation: jq -ec '.'
 
+  # Nesting Records in Lists is not currently fully supported, but its presence
+  # should also not crash the JSON reader.
+  Nested Records:
+    tags: [import, json]
+    steps:
+      - command: '-N import --schema="type custom = record { a: list<record { b: string, c: count, }>, d: count, }" --type=custom json'
```
Comment (dominiklohmann) [inline vast/integration/default_set.yaml:446]: This may only run if compiled with Arrow support.

File: libvast/src/arrow_table_slice.cpp (hunk 10 above)
Comment (mavam) [reply libvast/src/arrow_table_slice.cpp:267]: Yes please. If the type is known you should use `static_cast`.

**Pull request merged**